ethexe/processor/src/handling/run.rs

```diff
@@ -16,178 +16,259 @@
+//!
+//! Once all program queues have been processed, we deterministically merge journals and handle them.
+//! After that, we repeat the process until no more messages remain
+//! or we run out of processing time/gas allowance (to be implemented).
+//!
+//! **High-level overview of the algorithm:**
+//!
+//!   1. Split programs into buckets based on their queue sizes.
+//!   2. Execute program queues in parallel using runtime instances per program.
+//!   3. Merge journals and handle them deterministically.
+//!   4. Repeat steps 1-3 until no messages are processed, or we run out of processing time/gas allowance.
+//!
+//! ---
+//!
+//! ## Simplest Bucket Splitting Algorithm
+//!
+//! A basic bucket-splitting algorithm is implemented as follows:
+//!
+//!   1. First, calculate a temporary bucket index based on the program queue size.
+//!   2. Next, store the required data in a temporary bucket list.
+//!   3. Repeat this process for all programs.
+//!   4. Since the number of elements in each temporary bucket is random, they need to be redistributed
+//!      to ensure all final buckets contain an equal number of elements.
+//!      To achieve this, we first merge all temporary bucket lists sequentially
+//!      and then redistribute the data according to the expected bucket size.
+//!
+//! ---
+//!
+//! ## Future Improvements
+//!
+//! Currently, the bucket partitioning algorithm is simple and does not consider a program’s execution time.
+//! In the future, we could introduce a weight multiplier to the queue size to improve partitioning efficiency.
+//! This weight multiplier could be calculated based on program execution time statistics.
+
+use std::iter;
+
 use ethexe_db::{CodesStorage, Database};
-use ethexe_runtime_common::{InBlockTransitions, JournalHandler, TransitionController};
-use gear_core::ids::ProgramId;
-use gprimitives::H256;
-use std::collections::BTreeMap;
-use tokio::sync::{mpsc, oneshot};
-
-enum Task {
-    Run {
-        program_id: ProgramId,
-        state_hash: H256,
-        result_sender: oneshot::Sender<(Vec<JournalNote>, Option<Origin>)>,
-    },
-}
+use ethexe_runtime_common::{
+    InBlockTransitions, JournalHandler, ProgramJournals, TransitionController,
+};
+use gprimitives::{ActorId, H256};
+
+use crate::host::{InstanceCreator, InstanceWrapper};
 
-pub fn run(
-    config: &ProcessorConfig,
+pub async fn run(
+    queues_processing_threads: usize,
     db: Database,
     instance_creator: InstanceCreator,
     in_block_transitions: &mut InBlockTransitions,
 ) {
-    tokio::task::block_in_place(|| {
-        let mut rt_builder = tokio::runtime::Builder::new_multi_thread();
+    let mut join_set = tokio::task::JoinSet::new();
+    let max_bucket_size = queues_processing_threads;
 
-        if let Some(worker_threads) = config.worker_threads_override {
-            rt_builder.worker_threads(worker_threads);
-        };
+    loop {
+        let mut no_message_processed = true;
```

Comment:
1) The meaning of this variable is not "no messages processed" now
2) It's better to break after splitting to buckets like this
```
if buckets.is_empty() {
    break;
}
```

Reply:
Fixed